nmmo/core/observation.py

```diff
@@ -100,15 +118,193 @@ def to_gym(self):
+        action.Price: None # allow any integer
+      }
+      masks[action.Buy] = {
+        action.MarketItem: self._generate_buy_mask()
+      }
+
+    if self.config.COMMUNICATION_SYSTEM_ENABLED:
+      masks[action.Comm] = {
+        action.Token: self._generate_allow_all_mask(action.Token.edges),
+      }
+
+    return masks
+
+  def _generate_allow_all_mask(self, actions):
+    return np.ones(len(actions), dtype=np.int8)
+
+  def _generate_move_mask(self):
+    # pylint: disable=not-an-iterable
+    return np.array(
+      [self.tile(*d.delta).material_id in material.Habitable
+       for d in action.Direction.edges], dtype=np.int8)
+
+  def _generate_attack_mask(self):
+    # TODO: Currently, all attacks have the same range
+    #   if we choose to make ranges different, the masks
+    #   should be differently generated by attack styles
+    assert self.config.COMBAT_MELEE_REACH == self.config.COMBAT_RANGE_REACH
+    assert self.config.COMBAT_MELEE_REACH == self.config.COMBAT_MAGE_REACH
+    assert self.config.COMBAT_RANGE_REACH == self.config.COMBAT_RANGE_REACH
+
+    attack_range = self.config.COMBAT_MELEE_REACH
+
+    agent = self.agent()
+    dist_from_self = self.entities.dist((agent.row, agent.col))
+    not_same_tile = dist_from_self > 0 # this also includes not_self
+    within_range = dist_from_self <= attack_range
+
+    if not self.config.COMBAT_FRIENDLY_FIRE:
+      population = self.entities.values[:,EntityState.State.attr_name_to_col["population_id"]]
+      no_friendly_fire = population != agent.population_id
+    else:
+      # allow friendly fire
+      no_friendly_fire = np.ones(self.entities.len, dtype=np.int8)
+
+    return np.concatenate([not_same_tile & within_range & no_friendly_fire,
+      np.zeros(self.config.PLAYER_N_OBS - self.entities.len, dtype=np.int8)])
+
+  def _generate_use_mask(self):
+    # empty inventory -- nothing to use
+    if self.inventory.len == 0:
+      return np.zeros(self.config.INVENTORY_N_OBS, dtype=np.int8)
+
+    not_listed = self.inventory.values[:,ItemState.State.attr_name_to_col["listed_price"]] == 0
+    item_type = self.inventory.values[:,ItemState.State.attr_name_to_col["type_id"]]
+    item_level = self.inventory.values[:,ItemState.State.attr_name_to_col["level"]]
+
+    # level limits are differently applied depending on item types
+    type_flt = np.tile ( np.array(list(self._item_skill.keys())), (self.inventory.len,1) )
+    level_flt = np.tile ( np.array(list(self._item_skill.values())), (self.inventory.len,1) )
+    item_type = np.tile( np.transpose(np.atleast_2d(item_type)), (1, len(self._item_skill)))
+    item_level = np.tile( np.transpose(np.atleast_2d(item_level)), (1, len(self._item_skill)))
+    level_satisfied = np.any((item_type == type_flt) & (item_level <= level_flt), axis=1)
+
+    return np.concatenate([not_listed & level_satisfied,
+      np.zeros(self.config.INVENTORY_N_OBS - self.inventory.len, dtype=np.int8)])
+
+  @property
+  def _item_skill(self):
```

Comment:
i don't think this needs to be @property, but you should probably just compute it once in _generate_use_mask()

Reply:
deleted @property and just using it once in _make_use_mask()